### ISSUE TITLE
Fast GAN training in Keras

### ISSUE THREAD

**Issue by ctmakro** [fchollet/keras#5312]
To efficiently train a GAN, you must be able to update different set of weights w.r.t different losses, in one forward pass. That's why all GAN implemented in Keras are instead using the inefficient method(create two training model, and fit them separately).

As a fan of Keras, after digging thru TF documentation I wrote some code that could update different sets of weights from different models w.r.t. different losses in one forward-backward pass. That's at least 2X faster than the inefficient method.

Implementation detail and explainations: <https://ctmakro.github.io/site/on_learning/fast_gan_in_keras.html>

However I'm not sure what's the best way to embed this kind of functionality into Keras. Maybe a new `GANModel` class that takes two model(D and G) as input, then call its `fit()` method?

**Comment by ctmakro** [comment]
here's another guy who wrote the theano version: github.com/bstriner/keras-adversarial

**Comment by esube** [comment]
@ctmakro Due to the high level abstraction of Keras API, it's currently inefficient to train a GAN and its cousins in Keras as you pointed out. I have wrestled with this for a while. Your solution is a good start but I think the long term solution is not to abstract away this code into another high level API which is specific to GANs. Rather, because of the increasing popularity of GANs and their derivatives, it would be wise for keras to create separate training functions that also accept losses to update weights. In short, supply separate API for forward and backward with different flavors just like Torch. As Keras was originally developed to mimic torch in python, it would be fitting. 

@fchollet any ideas on this? I work on DCGANs extensively these days and I see that almost half the papers under review for ICLR 2017 are GAN related in some sense. Is there any interest to provide such APIs as forward and backward similar to torch in the core Keras?

**Comment by ctmakro** [comment]
@esube indeed. I recently wrote a lightweight wrapper library (<https://github.com/ctmakro/canton> ) around Tensorflow to do two things:

1. very intuitively creating and selecting weight tensors of a model
2. (therefore we can) update different weights wrt different losses very easily

It worked fine as long as you don't teach them to highschool students.

**Comment by shaform** [comment]
Recently I've read [BEGAN: Boundary Equilibrium Generative Adversarial Networks](https://arxiv.org/abs/1703.10717).
It claims that "In contrast to traditional GANs which require alternating training D and G, or pretraining D, our proposed method BEGAN requires neither to train stably. The objectives are trained in parallel".

And the so called "parallel" training is still updating different set of weights w.r.t different losses. Which seems to indicate that for traditional GANs, we should not update weights in parallel.

However I could not find any sources that state this clearly. But indeed few implementations try to update them in parallel.

**Comment by patrickmesana** [comment]
GANs are becoming so important, this thread should be top priority if Keras wants to stay in the game. I started to play with GANs and Keras, it's slow and BatchNorm is buggy.

**Comment by ctmakro** [comment]
@shaform as far as I know(from experiment), update in parallel does not affect the results.

@patrickmesana since this issue I invested a lot of effort into the training of GANs. By writing my own library (https://github.com/ctmakro/canton , with an API much like sonnet) models are built faster with less restrictions than that of Keras. The GAN part however stays pretty much the same.

My conclusion: there is no common interface for GAN. there are simply too many flavor of GANs out there, using all kinds of topology, fitting all kinds of objectives. 
My recommendation for now: If you need GAN, please construct the update ops in TF manually, as shown below.

here's how I train LSGAN, where `d` and `g` are two networks:
```py
    generated = g(noise)

    gscore = d(generated)
    rscore = d(real_data)

    dloss = tf.reduce_mean((gscore-0)**2 + (rscore-1)**2)
    gloss = tf.reduce_mean((gscore-1)**2)

    Adam = tf.train.AdamOptimizer

    lr,b1 = tf.Variable(1.2e-4),.5 # otherwise won't converge.
    optimizer = Adam(lr,beta1=b1)

    update_wd = optimizer.minimize(dloss,var_list=d.get_weights())
    update_wg = optimizer.minimize(gloss,var_list=g.get_weights())
```
(https://github.com/ctmakro/hellotensor/blob/master/lets_gan_canton.py)

**Comment by Phylliida** [comment]
Could something like this work?
```
class ForwardBackwardLayer(Layer):
    def __init__(self, output_dim, **kwargs):
        kwargs['trainable'] = False
        self.output_dim = output_dim
        self.using_backwards = False
        super(ForwardBackwardLayer, self).__init__(**kwargs)

    def build(self, input_shape):
        super(ForwardBackwardLayer, self).build(input_shape)
        
    def call(self, x):
        if self.using_backwards:
          return x[0]
        else:
          return x[1]
```

Example usage:

```
input = Input((3,))

backwardLayer = Activation("softmax")(input)
forwardLayer = Activation("linear")(input)

output = ForwardBackwardLayer(3)([forwardLayer, backwardLayer])

model = Model(input, output)
model.compile(optimizer=optimizers.Adam(), loss="mse")

# Make a temporary model that uses the backwards functions (for computing gradients). Because we already compiled our model this won't affect it
for layer in model.layers:
  if hasattr(layer, "using_backwards"):
    layer.using_backwards = True

inputB = Input((3,))
# When we do this it does the "call" methods again, giving this model our backward functions
outputB = model(inputB)
modelB = Model(inputB, outputB)
modelB.compile(optimizer=optimizers.Adam(), loss="mse")

# Set our original model's optimizer to this optimizer so it uses these updated gradients
# Because both models share the same weights, this optimizer will update them both. Looking at the code I'm pretty sure doing this makes the output from our model (in the forward pass) be passed into this optimizer which then uses the backward pass gradients
 
model.optimizer = modelB.optimizer
```

If this doesn't work I think something along these lines might

**Comment by hadaev8** [comment]
Still not fixed, right?
